feat: refactor validators and fields to support pydantic v2

xiaobo8204 opened the pull request:
## Description

This PR aims to update Pydantic to the latest v2.10.6, and switch to use the new model_validator/Field/field_validator in the new Pydantic version.

## Related Issue(s)

  - Fixes #929

## Checklist

- [Y] I've read the [CONTRIBUTING](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [Y] I've updated the documentation if applicable.
- [Y] I've added tests if applicable.
- [Y] @mentions of the person or team responsible for reviewing proposed changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Chores**
  * Updated Pydantic dependency to v2 across the project for improved compatibility with modern Python ecosystem standards.
  * Refactored internal validation and serialization patterns across multiple modules for enhanced code maintainability.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Upgrade to Pydantic v2: Update Validators and Packaging

- Refactored code to use `@model_validator` instead of `@root_validator`
- Updated `Field` usage to align with Pydantic v2 standards
- Modified validation logic to comply with the new API
- Updated `poetry.lock` and dependencies in `pyproject.toml`

Pouyanpi commented:
@xiaobo8204 thanks for your contribution to NeMo Guardrails 👍🏻 would you please sign all your commits following the [contributing guidelines](https://github.com/NVIDIA/NeMo-Guardrails/blob/08569c8d287672e282bc6abbf5ad6978dcaef969/CONTRIBUTING.md#submitting-your-work). You should see a Verified tag for all your commits.

Pouyanpi commented on log.txt:
please remove it (do not stage it or add it to .gitignore)

xiaobo8204 replied:
removed

Pouyanpi commented on .coverage.xiaobos-MacBook-Pro-2.local.14392.XotChdwx:
please remove it (do not stage it or add it to .gitignore)

xiaobo8204 replied:
removed

Pouyanpi commented on pyproject.toml:
```diff
@@ -65,7 +65,7 @@ langchain-community = ">=0.0.16,<0.4.0"
 lark = ">=1.1.7"
 nest-asyncio = ">=1.5.6,"
 prompt-toolkit = ">=3.0"
-pydantic = ">=1.10"
```
we should relax it a bit, probably `>= 2.0`

And I still need to see the reasons we were still supporting Pydantic 1.0, I'll get back to this later.

xiaobo8204 replied:
removed the extra log.text and hidden file, changed the version to >= 2.0, and signed the commit, thanks...

Pouyanpi replied:
and <3.0

Pouyanpi commented:
@xiaobo8204 Apologies for the delay in reviewing your PR.

The issue is that we currently don’t have proper tests for the code your PR affects. I’ll open a separate PR to add the necessary tests. Once that’s in place, we can rebase your branch and move forward from there. Thanks for your patience!

Pouyanpi commented on nemoguardrails/llm/providers/trtllm/llm.py:
```diff
@@ -61,12 +61,12 @@ class TRTLLM(BaseLLM):
     client: Any
     streaming: Optional[bool] = True
 
```
should be

```python

    @classmethod
    def validate_environment(cls, self) -> "TRTLLM":
        """Validate that python package exists in environment."""
        try:
            # instantiate and attach the client
            self.client = TritonClient(self.server_url)

        except ImportError as err:
            raise ImportError(
                "Could not import triton client python package. "
                "Please install it with `pip install tritonclient[all]`."
            ) from err
        return self

```

Pouyanpi commented on nemoguardrails/eval/models.py:
```diff
@@ -147,11 +147,11 @@ class EvalConfig(BaseModel):
 
-    @root_validator(pre=False, skip_on_failure=True)
-    def validate_policy_ids(cls, values: Any):
+    @model_validator(mode="after")
```
should be

```python
    @model_validator(mode="after")
    def validate_policy_ids(cls, self) -> "EvalConfig":
        """Validates the policy ids used in the interactions."""
        policy_ids = {policy.id for policy in self.policies}
        for interaction_set in self.interactions:
            for expected_output in interaction_set.expected_output:
                if expected_output.policy not in policy_ids:
                    raise ValueError(
                        f"Invalid policy id {expected_output.policy} used in interaction set."
                    )
            for policy_id in (
                interaction_set.include_policies + interaction_set.exclude_policies
            ):
                if policy_id not in policy_ids:
                    raise ValueError(
                        f"Invalid policy id {policy_id} used in interaction set."
                    )
        return self

```

Pouyanpi committed:
refactor(config): migrate remaining root_validator and validator to pydantic v2

Convert the model-existence and jailbreak-detection root validators to
model_validator(mode="before") and the models field validator to
field_validator, removing the deprecated pydantic v1 validator shims so the
config module emits no PydanticDeprecatedSince20 warnings.

Pouyanpi committed:
refactor: replace remaining pydantic v1 method calls with v2 equivalents

Migrate the deprecated .dict()/.parse_obj()/.json() model calls to
model_dump()/model_validate()/model_dump_json() across config, eval, server and
library modules, and drop the field_validator-as-validator alias in the Trend
Micro action. The only remaining .dict() call is the guarded model_dump
fallback in server/schemas/utils.py, which is intentional.

Pouyanpi committed:
build: raise pydantic floor to >=2.5 for Discriminator string-arg API

config.py uses Discriminator("engine"), whose string-argument form was added in
pydantic 2.5.0, so >=2.0 would fail to resolve correctly on 2.0-2.4.

Pouyanpi committed:
refactor(trend_micro): add @classmethod and None-guard to set_blocked validator

Align the before-validator with the v2 canonical form used elsewhere and avoid an
AttributeError when action is absent from the input.

Pouyanpi commented on pyproject.toml:
```diff
@@ -65,7 +65,7 @@ nest-asyncio = ">=1.5.6,"
 protobuf = ">=5.29.5"
 prompt-toolkit = ">=3.0"
-pydantic = ">=1.10"
+pydantic = ">=2.5,<3.0"
```
@greptileai those are optional and imported lazyily, it shouldn't concern the main deps

Pouyanpi committed:
fix(trtllm): give client field a default for pydantic v2

In pydantic v1 an `Any`-typed field with no default was implicitly optional; v2
treats it as required. Without a default, client field validation fails before the
mode="after" validate_environment validator can populate it, so TRTLLM(server_url=...)
always raised ValidationError. Default client to None to restore v1 behavior.

Pouyanpi commented on nemoguardrails/integrations/langchain/providers/trtllm/llm.py:
```diff
@@ -62,18 +62,17 @@ class TRTLLM(BaseLLM):
     client: Any
     streaming: Optional[bool] = True
```
Good catch, confirmed and fixed in 98382298d.

Pouyanpi commented on pyproject.toml:
```diff
@@ -65,7 +65,7 @@ nest-asyncio = ">=1.5.6,"
 protobuf = ">=5.29.5"
 prompt-toolkit = ">=3.0"
-pydantic = ">=1.10"
+pydantic = ">=2.5,<3.0"
```
those are optional and imported lazyily, it shouldn't concern the main dep

Pouyanpi committed:
fix(pydantic): refine trtllm client field and trend_micro set_blocked per review

- trtllm: mark the validator-populated client field exclude=True so the internal
  Triton connection object stays out of the model schema and serialization.
- trend_micro: move set_blocked from mode="before" to mode="after" so it runs on
  the validated instance; this removes the AttributeError when action is a
  non-string (now a clean ValidationError) and drops the now-redundant None guard.
  Also compare action to the literal "Block" directly instead of .lower().

Pouyanpi commented on nemoguardrails/integrations/langchain/providers/trtllm/llm.py:
```diff
@@ -62,18 +62,17 @@ class TRTLLM(BaseLLM):
-    def validate_environment(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+    @model_validator(mode="after")
+    def validate_environment(self) -> "TRTLLM":
         """Validate that python package exists in environment."""
         try:
-            values["client"] = TritonClient(values["server_url"])
+            self.client = TritonClient(self.server_url)
 
         except ImportError as err:
             raise ImportError(
                 "Could not import triton client python package. Please install it with `pip install tritonclient[all]`."
             ) from err
-        return values
+        return self
```
fixed in latest commit

Pouyanpi reviewed (approved):
Finally! Thanks for your contribution @xiaobo8204 . I made further changes so it fits current state of the repo.
